include/valijson/adapters/yaml_cpp_adapter.hpp

```diff
@@ -0,0 +1,697 @@
+     */
+    YamlCppObjectMemberIterator(const YAML::Node::const_iterator &itr)
+        : m_itr(itr)
+    {
+    }
+
+    /**
+     * @brief   Returns a YamlCppObjectMember that contains the key and
+     * value belonging to the object member identified by the iterator.
+     */
+    YamlCppObjectMember operator*() const
+    {
+        return YamlCppObjectMember(m_itr->first.as<std::string>(),
+                                   m_itr->second);
+    }
+
+    DerefProxy<YamlCppObjectMember> operator->() const
+    {
+        return DerefProxy<YamlCppObjectMember>(**this);
+    }
+
+    /**
+     * @brief   Compare this iterator with another iterator.
+     *
+     * Note that this directly compares the iterators, not the underlying
+     * values, and assumes that two identical iterators will point to the same
+     * underlying object.
+     *
+     * @param   other  Iterator to compare with
+     *
+     * @returns true if the underlying iterators are equal, false otherwise
+     */
+    bool operator==(const YamlCppObjectMemberIterator &other) const
+    {
+        return m_itr == other.m_itr;
+    }
+
+    bool operator!=(const YamlCppObjectMemberIterator &other) const
+    {
+        return !(m_itr == other.m_itr);
+    }
+
+    const YamlCppObjectMemberIterator &operator++()
+    {
+        m_itr++;
+
+        return *this;
+    }
+
+    YamlCppObjectMemberIterator operator++(int)
+    {
+        YamlCppObjectMemberIterator iterator_pre(m_itr);
+        ++(*this);
+        return iterator_pre;
+    }
+
+  private:
+    /// Iternal copy of the original YamlCpp iterator
+    YAML::Node::const_iterator m_itr;
+};
+
+/// Specialisation of the AdapterTraits template struct for YamlCppAdapter.
+template <> struct AdapterTraits<valijson::adapters::YamlCppAdapter>
+{
+    typedef YAML::Node DocumentType;
+
+    static std::string adapterName()
+    {
+        return "YamlCppAdapter";
+    }
+};
+
+inline bool YamlCppFrozenValue::equalTo(const Adapter &other, bool strict) const
+{
+    return YamlCppAdapter(m_value).equalTo(other, strict);
+}
+
+inline YamlCppArrayValueIterator YamlCppArray::begin() const
+{
+    return m_value.begin();
+}
+
+inline YamlCppArrayValueIterator YamlCppArray::end() const
+{
+    return m_value.end();
+}
+
+inline YamlCppObjectMemberIterator YamlCppObject::begin() const
+{
+    return m_value.begin();
+}
+
+inline YamlCppObjectMemberIterator YamlCppObject::end() const
+{
+    return m_value.end();
+}
+
+inline YamlCppObjectMemberIterator
+YamlCppObject::find(const std::string &propertyName) const
+{
+    YAML::Node result = m_value[propertyName];
+    if (!result.IsDefined())
+        return end();
+
+    // yaml-cpp does not offer an iterator-based lookup,
+    // so instead we create a new placeholder object and
+    // return an iterator to that container instead.
+    YAML::Node wrapper = YAML::Node(YAML::NodeType::Map);
+    wrapper[propertyName] = result;
+    return YamlCppObjectMemberIterator(wrapper.begin());
```

Comment:
This is still a bit hacky, but it's better than the previous strategy of iterating to find the node.

I don't know if it's important that the returned iterator be comparable to the iterator generated
when reaching the same field through for-each iteration (other than `end()`). This implementation
won't support that, and I also don't know that there is an easy way to do it.

(I'm thinking it's not that important because order of iteration isn't guaranteed anyway, so I'm
not sure under what circumstances the comparison would be important.)